src/services/suggestionDiagnostics.ts

```diff
@@ -156,7 +156,29 @@ namespace ts {
     }
 
     function isPromiseHandler(node: Node): node is CallExpression {
-        return isCallExpression(node) && (hasPropertyAccessExpressionWithName(node, "then") || hasPropertyAccessExpressionWithName(node, "catch"));
+        return isCallExpression(node) && (
+            hasPropertyAccessExpressionWithName(node, "then") && hasZeroOrOneNonNullArguments(node) ||
+            hasPropertyAccessExpressionWithName(node, "catch"));
+    }
+
+    function hasZeroOrOneNonNullArguments(node: CallExpression) {
+        if (node.arguments.length < 2) return true;
+        let seenArg = false;
+        for (const arg of node.arguments) {
+            switch (arg.kind) {
+                case SyntaxKind.NullKeyword:
+                    continue;
+                case SyntaxKind.Identifier:
+                    if ((arg as Identifier).escapedText === "undefined") {

```

Comment:
In Roslyn, we had the option asking for the constant value of an expression.  I don't know if that's possible in TS, but it would be more accurate (e.g. `const noCatch = undefined`).

Reply:
I have a type checker available, but the parent function has never used it. It has always worked purely on syntax, and I don’t see a compelling reason to break from that here. (FWIW I don’t think we have anything built in to say “get me the type of this if it’s easy.” If it’s an identifier, I can call `checker.resolveName` on it to get the referenced symbol, but 99% of the time it’s going to be a function or variable declaration. If it’s a variable declaration, then you need even more semantic work to get the type of the variable.)